File: problems/HackerRank/kangaroo.ts
function kangaroo(x1: number, v1: number, x2: number, v2: number): string {
  // x1 + times*v1 = x2 + times*v2

  if (v1 === v2 && x2 !== x1) return 'NO';

  const times = (x2 - x1) / (v1 - v2);

  if (times > 0 && Number.isInteger(times)) {
    return 'YES';
  }

  return 'NO';
}

console.log(kangaroo(1, 5, 2, 2));

File: problems/HackerRank/week4/kangaroo.py
def kangaroo(x1, v1, x2, v2):
    # Write your code here
    # x1 + zv1 = x2 + zv2
    # z(v1 - v2) = x2 - x1
    # z = (x2 - x1) / (v1 - v2)
    
    if v1 == v2:
        if x2 == x1:
            return 'YES'
        return 'NO'

    z = (x2 - x1) / (v1 - v2)
    
    if z.is_integer() and z > 0:
        return 'YES'
    return 'NO'